processing/src/main/java/io/druid/segment/DimensionHandlerUtils.java

```diff
@@ -172,10 +185,56 @@ private static ColumnValueSelector getColumnValueSelectorFromDimensionSpec(
   private static <ColumnSelectorStrategyClass extends ColumnSelectorStrategy> ColumnSelectorStrategyClass makeStrategy(
       ColumnSelectorStrategyFactory<ColumnSelectorStrategyClass> strategyFactory,
       String dimName,
-      ColumnCapabilities capabilities
+      ColumnCapabilities capabilities,
+      boolean hasExFn
   )
   {
-    capabilities = getEffectiveCapabilities(dimName, capabilities);
+    capabilities = getEffectiveCapabilities(dimName, capabilities, hasExFn);
     return strategyFactory.makeColumnSelectorStrategy(capabilities);
   }
+
+  private static final Pattern LONG_PAT = Pattern.compile("[-|+]?\\d+");
+
+  // TopNBinaryFn relies on this returning the same Long object when no conversion is needed
```

Comment:
I've removed this comment, I moved the type conversions to where DimExtractionTopNAlgorithm adds results to the result builder, so returning the same object is no longer strictly necessary (still returns the same object if type is the same though)